### DISCUSSION TITLE
Interactive Simulations with pluto for teaching

### DISCUSSION THREAD

**Discussion by dorn-gerhard** [fonsp/Pluto.jl#879] [Cool Pluto projects]
Hi,
I am currently designing a free massive open online course on **Bayesian probability theory** which is pixelart story based ("Captain Bayes and her adventures in probability theory") and would like to use pluto notebooks for a part of the course.
[Trailer](https://www.youtube.com/watch?v=gTWU6JFHxXg) to the course: https://www.youtube.com/watch?v=gTWU6JFHxXg

The idea is to provide the students with **interactive simulations** that help them to comprehend the concepts of probability theory. It's like a computational lab.

The basic requirements are:

- the input of the students is quite basic (sliders, buttons, textfields) 
- there is a good explanation (Markup language) introducing the experiment (e.g. also with a quiz of expectation before you start the experiment)
- the return is a visualization (e.g. animation with gif illustrating a random walk, histogram, ...)
- it should be easy for students to understand the simulation and to create additional interactive simulations for other students (which julia / pluto provides already :) 
- it should work per one mouse click in the web browser (e.g. using pluto on binder, remotely using mybinder or jupyterhub, docker server solution on campus)

Desired features (which I was not able to realize so far - mostly concerns binder and server settings):

- notebooks should be fast (e.g. all packages precompiled just waiting for the user input (e.g. moving of sliders, pressing a start button)
- many users can experiment with the notebook at the same time (e.g. using many instances of the same notebook on a docker server)
- users can only use the sliders, buttons, the rest of the code is restricted

**An example** is binomial distribution: You repeat n times a binary experiment with two outcomes (e.g. red and blue balls) with probabilites p and 1-p and count the number k of red balls. Then you repeat the whole thing N times and visualize the distribution as histogram and compare it with the binomial distribution (n choose k)

I created a pluto example illustrating the central limit theorem:
https://github.com/dorn-gerhard/Bayes_IMOOX/blob/master/random_dice.jl

I am looking for advice / suggestions / help to realize the desired features, maybe you can give me some inspiration and I can report about the progress :)

**Comment by fonsp** [comment]
That sounds great! We are actively working on this feature.

Just a quick question: have you tried http://pluto-on-binder.glitch.me ? I you have, then perhaps you can edit your question to explain what features you are still missing?

**Reply by dorn-gerhard** [reply]
Thank you for the reply :)
Your pluto-on-binder solution is great and works quite well. 
I will have a closer look to learn how to implement binder for my needs on a local server and how to allow many accesses to the same pluto notebook, that do not interfer. At the moment I can watch how the figures change when another person is working on the same notebook (via the same mybinder link)

Yesterday I just had a look at the solution on the computationalthinking website (https://computationalthinking.mit.edu/Spring21/week1/) with sliders and instant working, that is kind of the thing I had in mind for my project.

Though I noted, that editing of the cells seems not to possible, you can delete the code or individal symbols which in some cases could allow to write a different program but I am not sure if the code is executed when pressing shift+enter.

I think all I need is to learn how to properly set up mybinder or how to use the pure html version (as in the homepage mentioned) to realize the interactive simulations for my course.

**Reply by fonsp** [reply]
https://github.com/fonsp/pluto-static-render-test

Here is a (still secret!!) template repository with the github action set up to automatically generate HTML exports with a binder button. As this is all work-in-progress, be prepared to do a bit of troubleshooting yourself.

Having interactive sliders on the website (without waiting for binder to load) is a very difficult challenge, that we only started solving last week. Right now I don't have any accessible info available on how to set this up yourself, but if you are comfortable with digitalocean apps/heroku, then feel free to snoop around the mit repo.

**Reply by dorn-gerhard** [reply]
Wow, sounds very interesting. 
We managed to use **PlutoBindServer** to create a solution that looks promising.
The only problem we encountered was the caching / the functionality of multiple sliders that provided arguments for one function (I call them multiparameter functions)

Changing one slider does not trigger a check of the values on the other sliders needed in the multiparameter function if one set of parameters has been already calculated. We also tried to switch off caching, which did not work. 
I also wonder, whether the show code feature can be used to have a look on the code if wished.

I created a minimal example:
https://captain-bayes.github.io/bayes-imoox/notebooks/test_multi_bind.html

**Comment by VarLad** [comment]
Boy oh boy! I discovered something interesting for myself! Thank you and your team @dorn-gerhard for this wonderful MOOC!

**Reply by dorn-gerhard** [reply]
Thank you 👍 
There are more funny and interesting adventures to come :)

**Comment by RajDandekar** [comment]
@dorn-gerhard I just discovered this course and thought that it was brilliant :) I was wondering how you managed to host the notebooks and retain their interactivity? Did you use PlutoSlideServer with servers on AWS/ DigitalOcean etc?
